### PR TITLE
[Simple Analytics] Subscriber Change Over Time Chart & Mailchimp API Integration

### LINKED ISSUE #200: [Simple Analytics] Subscriber Change Over Time Chart & Mailchimp API Integration
### Description
Build a diverging bar chart showing new subscribers (positive, blue) and unsubscribes (negative, red) over time using the Mailchimp Activity API, with period totals and graceful 180-day limitation handling.

**Technical Context:**

- Mailchimp API `lists/{list_id}/activity` returns up to 180 days with `subs` and `unsubs` fields
- For ranges beyond 180 days, show informational message and display available data
- Same aggregation interval rules as Form Performance chart
- API errors must not break the rest of the analytics page

### Implementation Breakdown

**1. REST API/AJAX Endpoint**

- `GET /mailchimp/v1/analytics/subscriber-activity`:
  - Params: `list_id`, `date_from`, `date_to`; Permission: `manage_options`
  - Fetch from Mailchimp API `lists/{list_id}/activity`, cache in transient (~15min)
  - Filter to requested date range, aggregate by interval according to the date range
  - Return: `{ data: [{ label, new_subscribers, unsubscribes }], total_new, total_unsubs, net_change, limited: bool }`
  - Set `limited: true` if date range > 180 days

**2. Diverging Bar Chart**

- Chart.js bar chart with positive (blue) and negative (red) bars
- Tooltip: New Subscribers, Unsubscribes per bar
- "Totals for the selected date range" summary: Net change (e.g., "+47"), New Subscribers, Unsubscribes

**3. API Error & Limitation Handling**

- If `limited: true`, show info notice about 180-day limit
- If Mailchimp API fails, show error message in chart section only (other sections unaffected)
- Loading state while fetching API data

**4. Error & Empty States**

- **No data**: charts show "No data available for the selected date range" (copy to be reviewed before final approval) 
- **API unavailable**: Subscriber chart shows error (copy to be reviewed before final approval); other sections unaffected
- **No connected forms**: Message with link to Form Settings
- **Loading states**: Skeleton/spinner per section; sections load independently

### Acceptance Criteria

- Diverging bar chart shows new subscribers (blue, positive) and unsubscribes (red, negative)
- Data sourced from Mailchimp Activity API, aggregated by correct interval according to the date range
- Tooltips show both metrics per bar
- Period totals (net change, new, unsubs) display correctly
- Ranges > 180 days show informational message with available data
- API failures show error in chart section without breaking KPIs or Form Performance chart
- API responses are cached to reduce Mailchimp API calls
- Chart updates on filter change

### PR DESCRIPTION
**Pull request by iamdharmesh**
### Description of the Change
The PR adds a Subscriber change over time section on the Analytics admin page, a diverging **bar chart** (new subscribers vs unsubscribes) and a totals **doughnut** with net change, driven by the Mailchimp Marketing API `lists/{list_id}/activity` endpoint. The section loads on the existing `mailchimp-analytics-refresh` event so it stays independent from KPIs and Form Performance section failures in this request do not block other analytics UI.

Screenshot 2026-04-20 at 1 09 50 PM

#### Technical Details:
**Backend:**
- New **`Mailchimp_Subscriber_Activity`** class (`includes/class-mailchimp-subscriber-activity.php`): admin AJAX action **`mailchimp_sf_get_subscriber_activity`**, capability check (`MCSF_CAP_THRESHOLD`), nonce (`mailchimp_sf_analytics_admin_nonce`), and validated `list_id` / `date_from` / `date_to` (`Y-m-d`).
- Fetches raw activity (180-day API window), normalizes to daily series, then **filters** to the requested range intersected with the last **180 days** (site timezone). Returns **`limited: true`** when the selected span exceeds what Mailchimp can supply so the UI can show an info notice.
- **Bucket interval** follows the plan: daily (≤30d), weekly ISO Mon–Sun (31–90d), monthly (91d–1y).
- **Transient cache** (15 minutes) on successful raw responses only; key scoped by list (`mailchimp_sf_subscriber_activity_` + hash) so one cache entry serves many date-range requests.
- Registered and **`init()`**’d from **`mailchimp.php`** next to existing analytics classes.

**Frontend:**

- **`includes/admin/templates/analytics.php`**: new card markup for the section (title, date subtitle, notice, chart + totals, loading/error/empty states). Strings use the **`mailchimp`** text domain.
- **`assets/js/analytics.js`**: subscribes to filter refresh, POSTs to `admin-ajax.php`, renders **Chart.js** bar (diverging: positive new subs, negative unsubs) and doughnut for range totals; handles empty, error, and limited-range notice.
- **`assets/css/analytics.scss`** (and built **`dist/`** assets via **`npm run build`**): layout (chart + totals column), legend, notice/error/loading states, responsive behavior.

Closes #200

### How to test the Change
1. Go to **Mailchimp > Analytics**.
2. Ensure the **"Subscriber Change Over Time"** graph is displayed, including both bar and doughnut charts.
3. Verify that changing the list and date presets (e.g., 7 / 30 / 90 / 180 / 365 / custom) refreshes the graph, and that buckets and labels match the aggregation rules.
4. Select a longer date range (>180 days) and confirm that a limitation notice is shown and only ~180 days of data are displayed in the bars.
5. Ensure the section shows the correct loading state.
6. Confirm that an appropriate error message is displayed in case of API failure (e.g., disconnected account or invalid credentials).
7. Check responsiveness and cross-browser compatibility.

### Credits
Props @iamdharmesh @builtbytay April Domingo Cameron Campbell

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/mailchimp/wordpress/blob/develop/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: includes/class-mailchimp-subscriber-activity.php [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+		$api = mailchimp_sf_get_api();
+		if ( ! $api ) {
+			return new WP_Error(
+				'mailchimp_sf_not_connected',
+				esc_html__( 'Mailchimp account is not connected.', 'mailchimp' )
+			);
+		}
+
+		$response = $api->get( 'lists/' . rawurlencode( $list_id ) . '/activity', self::API_WINDOW_DAYS );
```
[the remark this answers is not in the document]
Comment (iamdharmesh) [reply includes/class-mailchimp-subscriber-activity.php:117]: it uses `count`

File: assets/js/analytics.js [hunk 2]
```diff
@@ -9,7 +9,12 @@
  */
 import { Datepicker } from 'vanillajs-datepicker';
 import 'vanillajs-datepicker/css/datepicker.css'; // eslint-disable-line import/no-unresolved
-import '../css/analytics.css';
+import '../css/analytics.scss';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
```
[the remark this answers is not in the document]
Comment (iamdharmesh) [reply assets/js/analytics.js:18]: PHP enqueue adds `wp-i18n` as a dependency.

File: includes/admin/templates/analytics.php [hunk 3]
```diff
@@ -100,21 +100,144 @@
 					</div>
 				</div>
 
-			<div class="mailchimp-sf-analytics-content" id="mailchimp-sf-analytics-content">
-			</div>
+				<section
+					class="mailchimp-sf-analytics-card mailchimp-sf-sa is-loading"
+					data-section="subscriber-activity"
+					aria-labelledby="mailchimp-sf-sa-title"
+				>
```
[the remark this answers is not in the document]
Comment (iamdharmesh) [reply includes/admin/templates/analytics.php:107]: That was just a placeholder.

File: includes/admin/templates/analytics.php [hunk 4]
```diff
@@ -100,21 +100,144 @@
...
+									<span
+										id="mailchimp-sf-sa-total-new"
+										class="mailchimp-sf-sa__legend-value"
+									>&mdash;</span>
+								</li>
+								<li class="mailchimp-sf-sa__legend-item is-unsub">
+									<span class="mailchimp-sf-sa__legend-swatch" aria-hidden="true"></span>
+									<span class="mailchimp-sf-sa__legend-label">
+										<?php esc_html_e( 'Unsubscribe', 'mailchimp' ); ?>
+									</span>
```
[the remark this answers is not in the document]
Comment (iamdharmesh) [reply includes/admin/templates/analytics.php]: Fixed now.

File: includes/class-mailchimp-subscriber-activity.php [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+			return 'weekly';
+		}
+		if ( $days <= 365 ) {
+			return 'monthly';
+		}
+		if ( $days <= 365 * 3 ) {
+			return 'quarterly';
+		}
+		return 'yearly';
+	}
```
[the remark this answers is not in the document]
Comment (iamdharmesh) [reply includes/class-mailchimp-subscriber-activity.php:252]: Yes. this will be used in other graph section.

File: includes/class-mailchimp-analytics.php [hunk 6]
```diff
@@ -90,11 +90,20 @@ public function enqueue_scripts( $hook_suffix ) {
...
+			$version      = $asset['version'] ?? MCSF_VER;
+		}
+
 		wp_enqueue_script(
 			'mailchimp_sf_analytics_js',
 			MCSF_URL . 'dist/js/analytics.js',
-			array( 'mailchimp_sf_chartjs' ),
-			MCSF_VER,
+			$dependencies,
+			$version,
```
[the remark this answers is not in the document]
Comment (iamdharmesh) [reply includes/class-mailchimp-analytics.php]: Good Point, fixed.

**Review by alaca — APPROVED** [review]
Looks good 👍

**Comment by ankitguptaindia** [comment]
## **QA Status:** Needs Feedback ❓

Hi @iamdharmesh,

Thanks for working on this—overall, it works as per the defined scope. I have a few questions and observations related to the chart data and UI:

1. **Data refresh interval:**
   Is there any specific time interval for updating the chart data?
   Currently, I have 3 subscribers in the Mailchimp list, but the chart is showing only 1. Is there a delay or sync interval for reflecting updated data in WordPress?

image

image


--------

2. **Unsubscribes count delay:**
   Similarly, the “Unsubscribes” count also seems to take time to update in the chart. Could you confirm if this is expected or we can improve this to show real time data?

--------

3. **UI issue – text clipping:**
   Text like “days” is getting cut off due to a fixed height.
   If `height: 36px;` is removed and set to auto, the UI appears correct. Can we adjust this to prevent character clipping (e.g., for letters like “y”)?

image


------

4. **Icon color consistency:**
   Can we align the colors for the “Subscribers” and “Unsubscribers” icons for better visual consistency?


image



**Next Step:** Moving back to In-Progress for feedback.

**Comment by iamdharmesh** [comment]
Thanks for testing this @ankitguptaindia.

> 1. Data refresh interval:
Is there any specific time interval for updating the chart data?

Yes, we are using transient caching which expires after every 15 mins. So, you will get updated data in 15 mins.

> 3. UI issue – text clipping:

Done. Thanks.

> 4. Icon color consistency:

Done. Thanks

**Comment by ankitguptaindia** [comment]
## QA Status: Approved ✅ 



## Test Summary

- Feature: “Subscriber Change Over Time” analytics section (diverging bar chart + totals doughnut) on Mailchimp > Analytics.
- **Status: Passed. Functionality works as expected in testing, and QA feedback has been addressed.**   


***

## Test Scenarios and Results

### 1. Basic rendering

- Opened **Mailchimp > Analytics**.  
- Verified the **“Subscriber Change Over Time”** card renders with: title, date subtitle, bar chart, doughnut chart, legend, totals, and loading/empty/error states.   
- Result: Pass. UI elements display correctly and use the `mailchimp` text domain.   ✅ 

### 2. List and date filters

- Changed list selector and presets: 7 / 30 / 90 / 180 / 365 days and custom ranges.   
- Verified charts refresh on `mailchimp-analytics-refresh`, bucket logic (daily ≤30d, weekly 31–90d, monthly 91d–1y), and labels match the aggregation rule.   
- Result: Pass. Bucketing and label behavior match the described technical details. ✅ 

### 3. 180‑day data limitation

- Selected a date range >180 days.   
- Verified that only ~180 days of data is shown, and the limitation notice appears.   
- Result: Pass. Mailchimp API window constraint is clearly communicated in the UI. ✅ 

### 4. Loading, error, and empty states

- Observed the loading state while data is fetched via `admin-ajax.php`.   
- Simulated failure (e.g., disconnected account) and verified that a page is not appearing.   
- Tested a list with no recent activity and confirmed the empty state is shown.   
- Result: Pass. All states behave as expected.    ✅ 

### 5. Data freshness and caching

- Verified updated subscriber and unsubscribe counts appear after cache expiry (expected delay up to 15 minutes).   
- Result: Pass. Behavior aligns with the documented 15‑minute refresh interval, and QA’s question on delayed counts is explained by this cache.   

### 6. UI/UX fixes from QA feedback

From QA feedback by `@ankitguptaindia`, the following items were raised and addressed:   

- Text clipping (“days” truncated due to fixed height):  
  - Fix applied (removal/adjustment of `height: 36px;` to prevent character clipping).   
  - Result: Pass. No clipping observed after the fix.    ✅ 
- Icon color consistency between “Subscribers” and “Unsubscribers”:  
  - Colors updated for consistent visual appearance.   
  - Result: Pass. Icons now use aligned colors.    ✅ 
- Chart color adjustment:  
  - Commit “Chart color fix” ensures diverging bar and doughnut colors follow the intended palette.   
  - Result: Pass. Colors are consistent and distinguishable.  ✅ 

QA feedback on these points is marked as **Done** by the author in the latest comment.   



### 6. Cross‑browser and responsive

- Checked the new card layout and charts for responsiveness and basic cross‑browser behaviour.
- Result: Pass. Layout remains usable and visually consistent at common breakpoints.    ✅ 


***

## Known/Expected Behaviours

- Data is not strictly real‑time due to the 15‑minute transient cache; subscribers and unsubscribers will update after cache expiry. This is expected behaviour, as confirmed by the author.   

***

## Final QA Status

- Overall Result: **Passed**.  
- QA feedback: Addressed (data freshness clarified; UI clipping fixed; icon and chart colors updated).   
- Recommendation: Ready to proceed with the merge. 


Screenshot 2026-04-30 at 9 16 30 PM

**Review by ankitguptaindia — APPROVED** [review]
